vlasovsolver/cpu_trans_pencils.cpp

```diff
@@ -0,0 +1,1170 @@
+               if (iSrc!=0 && mpiGrid.get_refinement_level(nbrPair.first)!=myRefLevel) {
+                  iSrc = -1;
+                  break;
+               }
+               // Flag as seed id if VLASOV_STENCIL_WIDTH positive neighbour is at higher refinement level
+               if (iSrc==0 && mpiGrid.get_refinement_level(nbrPair.first)>myRefLevel) {
+                  addToSeedIds = true;
+                  break;
+               }
+            }
+         }
+         iSrc--;
+      } // Finish B check
+
+      if ( addToSeedIds ) {
+         #pragma omp critical
+         seedIds.push_back(celli);
+         continue;
+      }
+      /* Proceed with C, checking if the next two negative neighbours have the same refinement level as ccell, but the
+         third neighbour a higher one. Iterate through negative distances for VLASOV_STENCIL_WIDTH+1 elements
+         starting from the smallest distance. */
+      iSrc = VLASOV_STENCIL_WIDTH;
+      for (auto it = distancesminus.begin(); it != distancesminus.end(); ++it) {
+         if (iSrc < 0) break; // found enough elements
+         for (const auto& nbrPair : *nbrPairs) {
+            int distanceInRefinedCells = -nbrPair.second[dimension];
+            if(distanceInRefinedCells == *it) {
+               // Break search if we are not at the final entry, and have different refinement level
+               if (iSrc!=0 && mpiGrid.get_refinement_level(nbrPair.first)!=myRefLevel) {
+                  iSrc = -1;
+                  break;
+               }
+               // Flag as seed id if VLASOV_STENCIL_WIDTH+1 positive neighbour is at higher refinement level
+               if (iSrc==0 && mpiGrid.get_refinement_level(nbrPair.first)>myRefLevel) {
+                  addToSeedIds = true;
+                  break;
+               }
+            }
+         }
+         iSrc--;
+      } // Finish C check
+
+      if ( addToSeedIds ) {
+#pragma omp critical
+         seedIds.push_back(celli);
+      }
+   }
+
+   if(debug) {
+      cout << "Rank " << myRank << ", Seed ids are: ";
+      for (const auto seedId : seedIds) {
+         cout << seedId << " ";
+      }
+      cout << endl;
+   }
+}
+
+/* Check whether the ghost cells around the pencil contain higher refinement than the pencil does.
+ * If they do, the pencil must be split to match the finest refined ghost cell.
+ *
+ * @param mpiGrid DCCRG grid object
+ * @param pencils Pencil data struct
+ * @param dimension Spatial dimension
+ */
+void check_ghost_cells(const dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGrid,
+                       setOfPencils& pencils,
+                       uint dimension) {
+
+   const bool debug = false;
+   int neighborhood = getNeighborhood(dimension,VLASOV_STENCIL_WIDTH);
+
+   int myRank;
+   if(debug) {
+      MPI_Comm_rank(MPI_COMM_WORLD,&myRank);
+   }
+
+   std::vector<CellID> pencilIdsToSplit;
+
+// Thread this loop here
```

Comment:
Need this comment?
(and scheduling? :P)

Reply:
Deleted the comment. This is also a lightweight but non-constant loop so dynamic is good.